File: 2018/day-14.rb
# frozen_string_literal: true

def find_next_10(after, rec_1, rec_2)
  recipes = []
  recipes_created = 0
  recipes.push(rec_1)
  recipes.push(rec_2)

  elf_1 = 0
  elf_2 = 1

  while recipes_created < after + 10
    sum = recipes[elf_1] + recipes[elf_2]
    if sum >= 10
      recipes.push(1)
      recipes_created += 1
    end
    recipes.push(sum % 10)
    recipes_created += 1

    elf_1 = (elf_1 + (recipes[elf_1] + 1)) % recipes.size
    elf_2 = (elf_2 + (recipes[elf_2] + 1)) % recipes.size

    # p recipes
  end
  recipes[after..(after + 9)].join
end

def calc_recipes(before, rec_1, rec_2)
  recipes = []
  recipes_created = 0
  recipes.push(rec_1)
  recipes.push(rec_2)

  elf_1 = 0
  elf_2 = 1

  target = before.to_s.chars.map(&:to_i)
  target_length = target.length

  loop do
    puts recipes_created if (recipes_created % 10_000).zero?
    sum = recipes[elf_1] + recipes[elf_2]
    if sum >= 10
      recipes.push(1)
      recipes_created += 1
      # puts recipes[recipes.size-6..recipes.size-1].join
      # puts before
      if (recipes.size > 5) && (recipes[-target_length..] == target)
        puts 'WINNER'
        return recipes.size - target_length
      end
    end
    recipes.push(sum % 10)
    recipes_created += 1

    elf_1 = (elf_1 + (recipes[elf_1] + 1)) % recipes.size
    elf_2 = (elf_2 + (recipes[elf_2] + 1)) % recipes.size

    # p recipes
    next unless recipes.size > 5

    # puts recipes[recipes.size-6..recipes.size-1].join
    # puts before
    if recipes[-target_length..] == target
      puts 'WINNER'
      return recipes.size - target_length
    end
  end
end
# input_2 = 01245 #5 # 18
input_2 = 59_414 # 2018
# input_2 = 894501

rec_1 = 3
rec_2 = 7
# part_1 = find_next_10(input, rec_1, rec_2)

puts input_2
part_2 = calc_recipes(input_2, rec_1, rec_2)

puts "answer: #{part_2}"

File: 2018/day_14.go
package main

import (
	"fmt"
	"strconv"
)

// calcRecipes mirrors the Ruby 'calc_recipes' function
func calcRecipes(before int, rec1 int, rec2 int) int {
	// Convert 'before' to a slice of integers
	beforeStr := strconv.Itoa(before)
	target := make([]int, len(beforeStr))
	for i, c := range beforeStr {
		target[i] = int(c - '0')
	}
	targetLength := len(target)

	// Initialize recipes with the first two recipes
	recipes := []int{rec1, rec2}

	// Initialize elf positions
	elf1 := 0
	elf2 := 1

	for {
		// Debugging: Print progress every 10,000 recipes
		if len(recipes)%10000 == 0 {
			fmt.Println(len(recipes))
		}

		// Calculate the sum of the current recipes
		sum := recipes[elf1] + recipes[elf2]

		// Add new recipes based on the sum
		if sum >= 10 {
			recipes = append(recipes, 1)
			// Check if the target appears after adding '1'
			if len(recipes) >= targetLength {
				match := true
				for i := 0; i < targetLength; i++ {
					if recipes[len(recipes)-targetLength+i] != target[i] {
						match = false
						break
					}
				}
				if match {
					fmt.Println("WINNER")
					return len(recipes) - targetLength
				}
			}
		}
		recipes = append(recipes, sum%10)
		// Check if the target appears after adding the last digit
		if len(recipes) >= targetLength {
			match := true
			for i := 0; i < targetLength; i++ {
				if recipes[len(recipes)-targetLength+i] != target[i] {
					match = false
					break
				}
			}
			if match {
				fmt.Println("WINNER")
				return len(recipes) - targetLength
			}
		}

		// Update elf positions
		elf1 = (elf1 + recipes[elf1] + 1) % len(recipes)
		elf2 = (elf2 + recipes[elf2] + 1) % len(recipes)
	}
}

func main() {
	// Define your input values
	// Uncomment the desired input
	//input2 := 51589  // Expected Output: 9
	//input2 := 1245   // Previously 01245, converted to 1245; Expected Output: 5
	//input2 := 92510  // Expected Output: 18
	//input2 := 59414  // Expected Output: 2018
	input2 := 894501 // Adjust as needed

	// Initial recipes
	rec1 := 3
	rec2 := 7

	// Display the target
	fmt.Printf("Target Sequence: %d\n", input2)

	// Calculate the position
	part2 := calcRecipes(input2, rec1, rec2)

	// Display the result
	fmt.Printf("Answer: %d\n", part2)
}
